drivers/at_cmd/at_cmd.c

```diff
@@ -289,12 +416,28 @@ int at_cmd_write(const char *const cmd,
 
 void at_cmd_set_notification_handler(at_cmd_handler_t handler)
 {
+	int rc;
 
 	k_sem_take(&cmd_pending, K_FOREVER);
+	rc = append_notif_handler(&handler_to_add, handler);
+	k_sem_give(&cmd_pending);
 
-	notification_handler = handler;
+	if (rc == 0 && k_sem_count_get(&list_pending) != 0) {
```

Comment:
I'm afraid that `k_sem_count_get(&list_pending) != 0` check is not enough for assuring thread-safety. In case two or more threads use `at_cmd_set_notification_handler`/`at_cmd_remove_notification_handler` APIs and one of them is preempted while inside `maintain_list()`, there's a risk of violating one of the lists (if another thread calls these APIs).

To assure thread-safety, you need to actually take the semaphore, otherwise, we're open for (very rare and hard to reproduce) failures (very often fatal).

Reply:
Up to now, I do not see any direction on how to make it "good", since there is no thread-safe link-list implementation at hand.

To go ahead, I see several possibilities:
1. Leave it as is, and keep the change un-merged.
2. Restrict the use case so that only appending of the list is allowed.
3. at_cmd_set_notification_handler/at_cmd_remove_notification_handler simply return -EBUSY when the dispatcher is being run. Up to now, these function calls only happen during initialization phase. It doesn't make sense to have them called during run-time.
4. Ignore mutual exclusion. As mentioned before, there is no use-case to add elements during runtime.